bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java

```diff
@@ -0,0 +1,82 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.pojo;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+final class PojoSpecializationHelper {
+
+    @SuppressWarnings("unchecked")
+    static <V> TypeData<V> specializeTypeData(final TypeData<V> typeData, final List<TypeData<?>> typeParameters,
+                                              final TypeParameterMap typeParameterMap) {
+        if (!typeParameterMap.hasTypeParameters() || typeParameters.isEmpty()) {
+            return typeData;
+        }
+
+        Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap = typeParameterMap.getPropertyToClassParamIndexMap();
+        Either<Integer, TypeParameterMap> classTypeParamRepresentsWholeField = propertyToClassParamIndexMap.get(-1);
+        if (classTypeParamRepresentsWholeField != null) {
+            Integer index = classTypeParamRepresentsWholeField.map(i -> i, e -> {
+                throw new IllegalStateException("Invalid state, the whole class cannot be represented by a subtype.");
+            });
+            return (TypeData<V>) typeParameters.get(index);
+        } else {
+            return getTypeData(typeData, typeParameters, propertyToClassParamIndexMap);
+        }
+    }
+
+    private static <V> TypeData<V> getTypeData(final TypeData<V> parentTypeData, final List<TypeData<?>> specializedTypeParameters,
+                                               final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap) {
+        List<TypeData<?>> typeParameters = new ArrayList<>(parentTypeData.getTypeParameters());
+        List<TypeData<?>> subTypeParameters = new ArrayList<>();
+
+        for (int i = 0; i < typeParameters.size(); i++) {
+            subTypeParameters.add(getTypeData(typeParameters.get(i), specializedTypeParameters, propertyToClassParamIndexMap, i));
+        }
+        return TypeData.builder(parentTypeData.getType()).addTypeParameters(subTypeParameters).build();
+    }
+
+    private static TypeData<?> getTypeData(final TypeData<?> parentTypeData, final List<TypeData<?>> specializedTypeParameters,
+                                           final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap,
+                                           final int index) {
+        if (!propertyToClassParamIndexMap.containsKey(index)) {
+            return parentTypeData;
+        }
+        return propertyToClassParamIndexMap.get(index).map(l -> {
+                    if (parentTypeData.getTypeParameters().isEmpty()) {
+                        // Represents the whole typeData
+                        return specializedTypeParameters.get(l);
```

Comment:
The `typeData` doesn't have any type parameters, so the specialization must represent the whole type.